staging/src/k8s.io/client-go/discovery/cached/disk/size_limited_disk_cache.go

```diff
@@ -0,0 +1,208 @@
+	if err == nil {
+		c.appendFileInfo(fileInfo{name: key, size: size, modTime: time.Now()})
+	} else {
+		klog.Errorf("Error disk cache entry (%s):  [%v]", key, err)
+	}
+}
+
+// Delete removes the response with key from the cache, updating the
+// caches fileInfos and the caches currentSize.
+func (c *SizeLimitedDiskCache) Delete(key string) {
+	key = keyToFilename(key)
+	c.diskv.Erase(key)
+	c.removeFileInfo(key)
+}
+
+// Passing the size of a file that is going to be added, evicts entries
+// from the cache until the current cache size plus the passed size is below
+// the maximum cache disk size. Can update the cache fileInfos and currentSize
+// if evictions occur.
+func (c *SizeLimitedDiskCache) trimDiskCache(size uint64) {
+	for ((c.currentSize + size) > c.maxSize) && (len(c.fileInfos) > 0) {
+		fi := c.fileInfos[0] // Evict oldest file
+		err := c.diskv.Erase(fi.name)
+		if err == nil {
+			c.currentSize -= fi.size
+			c.fileInfos = c.fileInfos[1:]
+		} else {
+			// If disk cache eviction fails, just exit.
+			klog.Errorf("Error trimming disk cache: [%v]\n", err)
+			break
+		}
+	}
+}
+
+// IMPORTANT: This function MUST be the same as
+// "github.com/gregjones/httpcache/diskcache.keyToFilename()" in
+// order to maintain compatibility with existing http disk caches.
+func keyToFilename(key string) string {
+	h := md5.New()
+	io.WriteString(h, key)
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+// NewSizeLimitedDiskCache returns a new Cache up to "maxSize" number
+// of total file bytes in the cache. Uses the provided Diskv as
+// underlying storage. Walks the entire tree of files in the cache.
+// Can cause evictions from the cache if the initial size of the cache
+// is beyond "maxSize".
+func NewSizeLimitedDiskCache(maxSize uint64, diskv *diskv.Diskv) *SizeLimitedDiskCache {
+	fileInfos := []fileInfo{}
+	currentSize := uint64(0)
+	err := filepath.Walk(diskv.BasePath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		fileSize := uint64(info.Size())
+		currentSize += fileSize
+		fi := fileInfo{
+			name:    info.Name(),
+			size:    fileSize,
+			modTime: info.ModTime(),
+		}
+		fileInfos = append(fileInfos, fi)
+		return nil
+	})
+	if err != nil {
+		klog.Errorf("Error during disk cache walk: [%v]\n", err)
+	}
+	sort.Sort(byFileModTime(fileInfos)) // Sort by ascending modTime
+	cache := &SizeLimitedDiskCache{
+		maxSize:     maxSize,
+		currentSize: currentSize,
+		fileInfos:   fileInfos,
+		diskv:       diskv,
+	}
+	cache.trimDiskCache(uint64(0))
+	return cache
+}
+
+// fileInfo stores relevant information for files in the disk cache.
+// This structure stores the size of each file to determine the
+// size of the entire disk cache, and it stores the modTime to
+// determine the oldest file to evict from the cache when necessary.
+// The name is the filename, which should be an MD5 hash (see
+// keyToFilename()).
+type fileInfo struct {
```

Comment:
All of the operations on `fileInfo` are O(n) which isn't ideal, but is probably also not super concerning in this case since `n` is small.